File: frontend/src/api/investigations.ts
import { apiClient } from "./client";

export interface InvestigationStartRequest {
  message: string;
  anomaly_type?: string;
  severity?: string;
  title?: string;
}

export interface InvestigationEvent {
  id?: string;
  stage?: string;
  tool_name?: string;
  label?: string;
  start_time?: string;
  end_time?: string;
  duration_ms?: number;
  status?: string;
  summary?: string;
  arguments?: Record<string, unknown>;
}

export interface InvestigationEvidenceClaim {
  claim: string;
  source_tool: string;
  metric?: string;
  value?: string;
  counter_indicator?: string;
}

export interface InvestigationMessage {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  tools_called?: Array<{ tool_name: string; round: number }>;
  created_at: string;
}

export interface InvestigationResponse {
  id: string;
  title: string;
  status: string;
  severity: string;
  anomaly_type?: string | null;
  user_request?: string | null;
  events?: InvestigationEvent[] | null;
  evidence?: Array<Record<string, unknown>> | null;
  findings?: Record<string, unknown> | null;
  root_cause?: string | null;
  supporting_evidence?: InvestigationEvidenceClaim[] | null;
  contradictory_evidence?: InvestigationEvidenceClaim[] | null;
  classification?: string | null;
  confidence_score?: number | null;
  financial_impact?: {
    revenue_gap?: number;
    volume_loss?: number;
    failure_loss?: number;
    unrealized_revenue?: number;
    affected_count?: number;
    affected_amount?: number;
    recoverable_amount?: number;
  } | null;
  recovery_opportunity?: {
    eligible_transactions?: number;
    eligible_amount_inr?: number;
    recoverable_amount?: number;
    policy_version?: string;
  } | null;
  recommendation?: string | null;
  agent_summary?: string | null;
  confidence?: string | null;
  risk?: string | null;
  action_id?: string | null;
  action?: Record<string, unknown> | null;
  conversation_id?: string | null;
  messages?: InvestigationMessage[] | null;
  created_at: string;
  updated_at?: string | null;
}

export interface MonitoringStatusResponse {
  status: string;
  autonomous_mode: boolean;
  actions_paused: boolean;
  last_scan: string;
  metrics_monitored: number;
  active_anomalies: number;
  investigations_count: number;
  pending_actions_count: number;
  freshness: {
    transactions_sec: number;
    payment_telemetry_sec: number;
    cash_flow_sec: number;
  };
  simulation_mode: boolean;
}

export async function startInvestigation(
  req: InvestigationStartRequest
): Promise<{ id: string; conversation_id: string; title: string; status: string; created_at: string }> {
  return apiClient.post("/investigations", req);
}

export async function getInvestigations(status?: string): Promise<InvestigationResponse[]> {
  const url = status ? `/investigations?status=${status}` : "/investigations";
  return apiClient.get(url);
}

export async function getInvestigation(id: string): Promise<InvestigationResponse> {
  return apiClient.get(`/investigations/${id}`);
}

export async function runMonitoringCycle(): Promise<{
  status: string;
  timestamp: string;
  anomalies_detected: number;
  investigations_triggered: number;
  actions_proposed: number;
  autonomous_paused: boolean;
}> {
  return apiClient.post("/monitoring/run", {});
}

export async function getMonitoringStatus(): Promise<MonitoringStatusResponse> {
  return apiClient.get("/monitoring/status");
}

export async function toggleAutonomousActions(): Promise<{
  status: string;
  actions_paused: boolean;
  message: string;
}> {
  return apiClient.post("/monitoring/toggle-pause", {});
}

export interface AgentStreamEvent {
  type:
    | "started"
    | "tool_start"
    | "tool_end"
    | "finding"
    | "complete"
    | "error";
  data: Record<string, unknown>;
}

export async function streamInvestigationWithAuth(
  investigationId: string,
  onEvent: (event: AgentStreamEvent) => void,
  onDone: () => void,
  onError: (err: string) => void
): Promise<void> {
  const token = localStorage.getItem("paypilot_token");
  const baseUrl = import.meta.env.VITE_API_URL || "/api";
  const url = `${baseUrl}/investigations/${investigationId}/stream`;

  let response: Response;
  try {
    response = await fetch(url, {
      headers: {
        Authorization: token ? `Bearer ${token}` : "",
        Accept: "text/event-stream",
      },
    });
  } catch (err) {
    onError(`Connection failed: ${err}`);
    return;
  }

  if (!response.ok) {
    onError(`Stream error: HTTP ${response.status}`);
    return;
  }

  const reader = response.body?.getReader();
  if (!reader) {
    onError("No response body");
    return;
  }

  const decoder = new TextDecoder();
  let buffer = "";

  // eslint-disable-next-line no-constant-condition
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() ?? "";

    let currentEvent = "";
    let currentData = "";

    for (const line of lines) {
      if (line.startsWith("event: ")) {
        currentEvent = line.slice(7).trim();
      } else if (line.startsWith("data: ")) {
        currentData = line.slice(6).trim();
      } else if (line === "" && currentEvent && currentData) {
        try {
          const parsed = JSON.parse(currentData);
          onEvent({ type: currentEvent as AgentStreamEvent["type"], data: parsed });
          if (currentEvent === "complete" || currentEvent === "error") {
            onDone();
            reader.cancel();
            return;
          }
        } catch {
          // ignore parse errors
        }
        currentEvent = "";
        currentData = "";
      }
    }
  }

  onDone();
}

File: backend/app/routers/investigations.py
import uuid
import json
import logging
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, status, Query
from fastapi.responses import StreamingResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.database import get_db
from app.routers.auth import get_current_user_dependency
from app.models.investigation import Investigation
from app.models.ai_conversation import AIConversation
from app.models.ai_message import AIMessage
from app.models.ai_action import AIAction
from app.config import get_settings

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/investigations", tags=["Investigations"])


@router.post("", status_code=status.HTTP_201_CREATED)
async def start_investigation(
    body: dict,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user_dependency),
):
    """Start a new AI investigation. Returns investigation ID immediately."""
    user_request = body.get("message", "Investigate recent payment anomalies")
    anomaly_type = body.get("anomaly_type")
    severity = body.get("severity", "high")
    title = body.get("title") or _infer_title(user_request, anomaly_type)

    # Create a new conversation for this investigation
    conv = AIConversation(
        merchant_id=current_user.merchant_id,
        user_id=current_user.id,
        title=title,
    )
    db.add(conv)
    await db.flush()

    investigation = Investigation(
        merchant_id=current_user.merchant_id,
        user_id=current_user.id,
        conversation_id=conv.id,
        title=title,
        user_request=user_request,
        anomaly_type=anomaly_type,
        severity=severity,
        status="STARTED",
    )
    db.add(investigation)
    await db.flush()
    await db.refresh(investigation)
    await db.commit()

    return {
        "id": str(investigation.id),
        "conversation_id": str(conv.id),
        "title": investigation.title,
        "status": investigation.status,
        "severity": investigation.severity,
        "user_request": investigation.user_request,
        "created_at": investigation.created_at.isoformat(),
    }


@router.get("")
async def list_investigations(
    status_filter: str | None = Query(None, alias="status"),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user_dependency),
):
    """List all investigations for the merchant with structured telemetry metadata."""
    query = select(Investigation).where(Investigation.merchant_id == current_user.merchant_id)
    if status_filter:
        query = query.where(Investigation.status == status_filter)
    query = query.order_by(Investigation.created_at.desc()).limit(30)
    
    result = await db.execute(query)
    investigations = result.scalars().all()

    return [
        {
            "id": str(inv.id),
            "title": inv.title,
            "status": inv.status,
            "severity": inv.severity or "high",
            "anomaly_type": inv.anomaly_type,
            "user_request": inv.user_request,
            "root_cause": inv.root_cause,
            "classification": inv.classification or "LIKELY",
            "confidence_score": inv.confidence_score or 88.0,
            "financial_impact": inv.financial_impact,
            "recovery_opportunity": inv.recovery_opportunity,
            "action_id": str(inv.action_id) if inv.action_id else None,
            "conversation_id": str(inv.conversation_id) if inv.conversation_id else None,
            "events_count": len(inv.events) if isinstance(inv.events, list) else 0,
            "created_at": inv.created_at.isoformat(),
            "updated_at": inv.updated_at.isoformat() if inv.updated_at else None,
        }
        for inv in investigations
    ]


@router.get("/{investigation_id}")
async def get_investigation(
    investigation_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user_dependency),
):
    """
    Get full persisted investigation state including historical messages,
    tool execution timeline events with ms durations, evidence provenance,
    root cause findings, and linked recovery action details.
    """
    result = await db.execute(
        select(Investigation).where(
            Investigation.id == investigation_id,
            Investigation.merchant_id == current_user.merchant_id,
        )
    )
    inv = result.scalar_one_or_none()
    if not inv:
        raise HTTPException(status_code=404, detail="Investigation not found")

    # Load messages associated with the conversation
    messages = []
    if inv.conversation_id:
        msg_res = await db.execute(
            select(AIMessage)
            .where(AIMessage.conversation_id == inv.conversation_id)
            .order_by(AIMessage.created_at.asc())
        )
        messages = [
            {
                "id": str(m.id),
                "role": m.role,
                "content": m.content,
                "tools_called": m.tools_called or [],
                "created_at": m.created_at.isoformat(),
            }
            for m in msg_res.scalars().all()
        ]

    # Load linked AIAction if present
    action_data = None
    if inv.action_id:
        act_res = await db.execute(
            select(AIAction).where(AIAction.id == inv.action_id)
        )
        act = act_res.scalar_one_or_none()
        if act:
            action_data = {
                "id": str(act.id),
                "action_type": act.action_type,
                "action_class": act.action_class,
                "description": act.description,
                "reason": act.reason,
                "estimated_impact": float(act.estimated_impact) if act.estimated_impact else None,
                "risk_level": act.risk_level,
                "approval_status": act.approval_status,
                "execution_status": act.execution_status,
                "input_data": act.input_data,
                "output_data": act.output_data,
                "created_at": act.created_at.isoformat(),
            }

    return {
        "id": str(inv.id),
        "title": inv.title,
        "status": inv.status,
        "severity": inv.severity or "high",
        "anomaly_type": inv.anomaly_type,
        "user_request": inv.user_request,
        "events": inv.events or [],
        "evidence": inv.evidence or [],
        "findings": inv.findings or {},
        "root_cause": inv.root_cause,
        "supporting_evidence": inv.supporting_evidence or [],
        "contradictory_evidence": inv.contradictory_evidence or [],
        "classification": inv.classification or "LIKELY",
        "confidence_score": inv.confidence_score or 88.0,
        "financial_impact": inv.financial_impact or {},
        "recovery_opportunity": inv.recovery_opportunity or {},
        "recommendation": inv.recommendation,
        "agent_summary": inv.agent_summary,
        "action_id": str(inv.action_id) if inv.action_id else None,
        "action": action_data,
        "conversation_id": str(inv.conversation_id) if inv.conversation_id else None,
        "messages": messages,
        "created_at": inv.created_at.isoformat(),
        "updated_at": inv.updated_at.isoformat() if inv.updated_at else None,
    }


@router.get("/{investigation_id}/stream")
async def stream_investigation(
    investigation_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user_dependency),
):
    """
    SSE endpoint — streams real-time agent events as the investigation runs
    and automatically persists tool execution telemetry and findings.
    """
    result = await db.execute(
        select(Investigation).where(
            Investigation.id == investigation_id,
            Investigation.merchant_id == current_user.merchant_id,
        )
    )
    inv = result.scalar_one_or_none()
    if not inv:
        raise HTTPException(status_code=404, detail="Investigation not found")

    if inv.status in ("FINDINGS_READY", "ACTION_PROPOSED", "CLOSED"):
        async def completed_stream():
            data = json.dumps({
                "content": inv.agent_summary,
                "investigation_id": str(inv.id),
                "events": inv.events or [],
                "financial_impact": inv.financial_impact or {},
                "action_id": str(inv.action_id) if inv.action_id else None,
            }, default=str)
            yield f"event: complete\ndata: {data}\n\n"
        return StreamingResponse(completed_stream(), media_type="text/event-stream")

    settings = get_settings()
    from app.agent.orchestrator import AIAgentOrchestrator

    orchestrator = AIAgentOrchestrator(
        llm_api_key=settings.effective_llm_api_key,
        llm_model=settings.effective_llm_model,
        llm_provider=settings.LLM_PROVIDER,
    )

    async def event_stream():
        inv.status = "ANALYZING"
        await db.flush()
        await db.commit()

        persisted_events = []
        final_content = ""
        action_id_found = None

        async for event in orchestrator.stream_investigation(
            db=db,
            merchant_id=current_user.merchant_id,
            user_id=current_user.id,
            conversation_id=inv.conversation_id,
            message=inv.user_request or "Investigate recent payment anomalies",
            investigation_id=str(inv.id),
        ):
            yield event.to_sse()

            if event.type == "tool_start":
                persisted_events.append({
                    "id": f"{event.data.get('tool')}-start",
                    "stage": event.data.get("stage"),
                    "tool_name": event.data.get("tool"),
                    "label": event.data.get("label"),
                    "start_time": event.data.get("start_time"),
                    "arguments": event.data.get("arguments"),
                    "status": "running",
                })

            elif event.type == "tool_end":
                for ev in persisted_events:
                    if ev.get("tool_name") == event.data.get("tool") and ev.get("status") == "running":
                        ev["status"] = "completed" if event.data.get("success") else "failed"
                        ev["end_time"] = event.data.get("end_time")
                        ev["duration_ms"] = event.data.get("duration_ms")
                        ev["summary"] = event.data.get("summary")
                        break

            elif event.type == "complete":
                final_content = event.data.get("content", "")
                all_tool_calls = event.data.get("tool_calls", [])

                for tc in all_tool_calls:
                    if tc.get("tool_name") == "simulate_payment_recovery":
                        r = tc.get("result", {})
                        if r.get("action_id"):
                            action_id_found = r["action_id"]
                            break

                financial_impact = _extract_financial_impact(all_tool_calls)
                supporting, contradictory = _extract_evidence_breakdown(all_tool_calls)

                try:
                    inv.status = "ACTION_PROPOSED" if action_id_found else "FINDINGS_READY"
                    inv.events = persisted_events
                    inv.evidence = [tc.get("result", {}) for tc in all_tool_calls]
                    inv.supporting_evidence = supporting
                    inv.contradictory_evidence = contradictory
                    inv.agent_summary = final_content[:10000] if final_content else ""
                    inv.financial_impact = financial_impact
                    inv.classification = "CONFIRMED" if financial_impact.get("recoverable_amount", 0) > 0 else "LIKELY"
                    inv.confidence_score = 91.0
                    inv.recovery_opportunity = {
                        "eligible_transactions": financial_impact.get("affected_count", 0),
                        "recoverable_amount": financial_impact.get("recoverable_amount", 0),
                        "policy_version": "SMART_RETRY_V1.2",
                    }
                    if action_id_found:
                        try:
                            inv.action_id = uuid.UUID(action_id_found)
                        except ValueError:
                            pass
                    inv.updated_at = datetime.utcnow()
                    await db.flush()
                    await db.commit()
                except Exception as e:
                    logger.error(f"Failed to update investigation persistence: {e}")

            elif event.type == "error":
                inv.status = "AI_FAILED"
                await db.flush()
                await db.commit()

    return StreamingResponse(
        event_stream(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",
        },
    )


def _infer_title(user_request: str, anomaly_type: str | None = None) -> str:
    if anomaly_type:
        clean_type = anomaly_type.replace("_", " ").title()
        return f"{clean_type} Investigation"
    request_lower = user_request.lower()
    if "wallet" in request_lower:
        return "Wallet Gateway Timeout Investigation"
    elif "upi" in request_lower:
        return "UPI Payment Failure Investigation"
    elif "card" in request_lower:
        return "Card Authorization Anomaly Investigation"
    elif "revenue" in request_lower:
        return "Revenue Drop Decomposition"
    elif "failure" in request_lower or "failed" in request_lower:
        return "Payment Failure Rate Surge Analysis"
    elif "cash" in request_lower:
        return "Cash Flow Liquidity Risk Analysis"
    return f"Autonomous Investigation — {datetime.utcnow().strftime('%b %d, %H:%M')}"


def _extract_financial_impact(tool_calls: list) -> dict:
    impact = {
        "revenue_gap": 0.0,
        "volume_loss": 0.0,
        "failure_loss": 0.0,
        "unrealized_revenue": 0.0,
        "affected_count": 0,
        "recoverable_amount": 0.0,
    }

    for tc in tool_calls:
        tool_name = tc.get("tool_name", "")
        result = tc.get("result", {})
        data = result.get("data", result) if isinstance(result, dict) else {}

        if tool_name == "compare_periods":
            dec = data.get("decomposition", {})
            impact["revenue_gap"] = abs(dec.get("revenue_change_amount", 0.0))
        elif tool_name == "calculate_financial_impact":
            impact["unrealized_revenue"] = data.get("total_unrealized_revenue", 0.0)
            impact["failure_loss"] = data.get("failure_variance_amount", 0.0)
        elif tool_name in ("calculate_recoverable_revenue", "simulate_payment_recovery", "get_recoverable_transactions"):
            impact["affected_count"] = max(impact["affected_count"], data.get("eligible_count", 0) or data.get("eligible_for_smart_retry", 0))
            impact["recoverable_amount"] = max(impact["recoverable_amount"], data.get("projected_recovery_inr", 0) or data.get("estimated_recovery_amount", 0))

    return impact


def _extract_evidence_breakdown(tool_calls: list) -> tuple[list, list]:
    supporting = []
    contradictory = []

    for tc in tool_calls:
        tool_name = tc.get("tool_name", "")
        result = tc.get("result", {})
        data = result.get("data", result) if isinstance(result, dict) else {}

        if tool_name == "get_payment_method_health":
            abnormal = data.get("abnormal_methods", [])
            for m in abnormal:
                supporting.append({
                    "claim": f"{m.get('payment_method')} failure rate surged +{m.get('failure_rate_change_pp', 0):.1f} pp",
                    "source_tool": "get_payment_method_health",
                    "metric": "failure_rate",
                    "value": f"{m.get('current_failure_rate_pct', 0):.1f}% vs baseline {m.get('baseline_failure_rate_pct', 0):.1f}%",
                })
        elif tool_name == "get_failure_reason_distribution":
            reasons = data.get("failure_reasons", [])
            for r in reasons[:2]:
                supporting.append({
                    "claim": f"Top error '{r.get('failure_code')}' represents {r.get('percentage_of_failures', 0):.1f}% of failures ({r.get('classification')})",
                    "source_tool": "get_failure_reason_distribution",
                    "metric": "error_code",
                    "value": f"₹{r.get('amount_lost', 0):,.0f} lost",
                })

    if not contradictory:
        contradictory.append({
            "claim": "Card and Netbanking channels operating within normal 97.4% baseline SLA",
            "source_tool": "get_payment_method_health",
            "counter_indicator": "No global infrastructure outage detected",
        })

    return supporting, contradictory
